### PR TITLE
[SYCL][ESIMD] Fix compilation break occurring when bfloat16 constructor is used in a kernel

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

**Review by sarnex — COMMENTED** [review]
Sorry for the late review, I missed this somehow.

Can you bring me up to speed on the root cause and the solution? I looked at the internal bug tracker but didn't totally understand how it's related to this PR.

My understanding is that all `__devicelib_ConvertFToBF16INTEL` does is call ` __spirv_ConvertFToBF16INTEL`, at least that's what [this ](https://github.com/intel/llvm/blob/74b3d45ccbe63a6e3a8df7eb046a58328ae758b7/libdevice/bfloat16_wrapper.cpp#L17 )code seems to be doing.

Do we not link against whatever sycl device library contains the implementation for `__devicelib_ConvertFToBF16INTEL` so we never see the defn or something?

Thanks

**Comment by sarnex** [comment]
Thanks for the explanation.

I investigated it a bit more and it seems `__devicelib_ConvertBF16ToFINTEL` is never translated to `__spirv_ConvertFToBF16INTEL` even in working non-ESIMD cases, even the GPU VISA does a function call for it, so I guess it's supposed to live somewhere. 

It seems like we need to handle this on our side, so this solution makes sense, will review it in depth soon.

**Review by sarnex — APPROVED** [review]
this looks good, thanks. i dont understand why the bf16 feature was implemented this way but this looks to be a good way to deal with it.

**Review by v-klochkov — APPROVED** [review]

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 1]
```diff
@@ -703,6 +707,28 @@ static const ESIMDIntrinDesc &getIntrinDesc(StringRef SrcSpelling) {
...
+      .Case("__devicelib_ConvertBF16ToFINTEL", true)
+      .Default(false);
+}
+
+// Mangle deviceLib function to make it pass through the regular workflow
+// These functions are defined as extern "C" which Demangler that is used
+// fails to handle properly.
+static std::string mangleDevicelibFunction(StringRef FunctionName) {
+  if (isDevicelibFunction(FunctionName)) {
+    if (FunctionName.startswith("__devicelib_ConvertFToBF16INTEL")) {
```
[the remark this answers is not in the document]
Comment (sarnex) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:722]: ah cool idea, thanks

**Pull request merged**